atom/browser/net/atom_url_loader_factory.cc

```diff
@@ -243,25 +304,42 @@ void AtomURLLoaderFactory::SendResponseHttp(
       std::move(client), traffic_annotation);
 }
 
-bool AtomURLLoaderFactory::HandleError(
-    network::mojom::URLLoaderClientPtr* client,
-    v8::Isolate* isolate,
-    v8::Local<v8::Value> response) {
-  if (!response->IsObject())
-    return false;
-  v8::Local<v8::Object> obj =
-      response->ToObject(isolate->GetCurrentContext()).ToLocalChecked();
-  network::URLLoaderCompletionStatus status;
-  if (!mate::Dictionary(isolate, obj).Get("error", &status.error_code))
-    return false;
-  std::move(*client)->OnComplete(status);
-  return true;
+// static
+void AtomURLLoaderFactory::StartLoadingStream(
+    network::mojom::URLLoaderClientPtr client,
+    const mate::Dictionary& dict) {
+  network::ResourceResponseHead head = ToResponseHead(dict);
+  v8::Local<v8::Value> stream;
+  if (!dict.Get("data", &stream)) {
+    // Assume the opts is already a stream.
+    stream = dict.GetHandle();
+  } else if (stream->IsNullOrUndefined()) {
+    // "data" was explicitly passed as null or undefined, assume the user wants
+    // to send an empty body.
+    client->OnReceiveResponse(head);
+    client->OnComplete(network::URLLoaderCompletionStatus(net::OK));
+    return;
+  } else if (!stream->IsObject()) {
+    client->OnComplete(network::URLLoaderCompletionStatus(net::ERR_FAILED));
+    return;
+  }
+
+  mate::Dictionary data = ToDict(dict.isolate(), stream);
+  v8::Local<v8::Value> method;
+  if (!data.Get("on", &method) || !method->IsFunction() ||
+      !data.Get("removeListener", &method) || !method->IsFunction()) {
+    client->OnComplete(network::URLLoaderCompletionStatus(net::ERR_FAILED));
+    return;
+  }
+
+  new NodeStreamLoader(std::move(head), std::move(client), data.isolate(),
```

Comment:
Is it okay for this class to manage its own lifetime ?

Reply:
This seems to be the only choice, I couldn't find a way to have Chromium manage its lifetime. `content::FileURLLoader` is also managing its own lifetime:

https://cs.chromium.org/chromium/src/content/browser/file_url_loader_factory.cc?type=cs&g=0&l=341

Reply:
Oh cool, in that case should we check if `producer_` and `client_` handles are bound before accessing the methods when we receive data from JS land ?

Reply:
`content::FileURLLoader` does not check bound for every call but it deletes itself when connection error happens, I have added the same for `NodeStreamLoader`.